File: ios/Classes/Map_handler.swift
import Flutter
import UIKit
import MapKit
enum MapType: String {
    case apple
    case google
    case amap
    case baidu
    case waze
    case yandexNavi
    case yandexMaps
    
    func type() -> String {
        return self.rawValue
    }
}

class Map {
    let mapName: String
    let mapType: MapType
    let urlPrefix: String?
    
    init(mapName: String, mapType: MapType, urlPrefix: String?) {
        self.mapName = mapName
        self.mapType = mapType
        self.urlPrefix = urlPrefix
    }
    
    func toMap() -> [String:String] {
        return [
            "mapName": mapName,
            "mapType": mapType.type(),
        ]
    }
}

public class MapHandler: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "maps_channel", binaryMessenger: registrar.messenger())
        let instance = MapHandler()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    let maps: [Map] = [
        Map(mapName: "Apple Maps", mapType: MapType.apple, urlPrefix: ""),
        Map(mapName: "Google Maps", mapType: MapType.google, urlPrefix: "comgooglemaps://"),
        Map(mapName: "Amap", mapType: MapType.amap, urlPrefix: "iosamap://"),
        Map(mapName: "Baidu Maps", mapType: MapType.baidu, urlPrefix: "baidumap://"),
        Map(mapName: "Waze", mapType: MapType.waze, urlPrefix: "waze://"),
        Map(mapName: "Yandex Navigator", mapType: MapType.yandexNavi, urlPrefix: "yandexnavi://"),
        Map(mapName: "Yandex Maps", mapType: MapType.yandexMaps, urlPrefix: "yandexmaps://")
    ]
    
    func getMapByRawMapType(type: String) -> Map {
        return maps.first(where: { $0.mapType.type() == type })!
    }
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInstalledMaps":
            result(maps.filter({ isMapAvailable(map: $0) }).map({ $0.toMap() }))
        case "launchMap":
            guard let args = call.arguments as? [String: String],
                let mapType = args["mapType"],
                let url = args["url"],
                let title = args["title"],
                let address = args["address"] else { return }
            
            let map = getMapByRawMapType(type: mapType ?? "")
            if (!isMapAvailable(map: map)) {
                result(FlutterError(code: "MAP_NOT_AVAILABLE", message: "Map is not installed on a device", details: nil))
                return;
            }
            
            launchMap(mapType: MapType(rawValue: mapType)!, url: url, title: title)
        case "isMapAvailable":
            guard let args = call.arguments as? [String: String],
                let mapType = args["mapType"] else { return }
            let map = getMapByRawMapType(type: mapType)
            result(isMapAvailable(map: map))
        default:
            print("method does not exist")
        }
    }

    func launchMap(mapType: MapType, url: String, title: String) {
        //For now open url for all the maps because we are using address instead of coordinates.
        UIApplication.shared.openURL(URL(string:url)!)
    }
    
    func isMapAvailable(map: Map) -> Bool {
        if map.mapType == MapType.apple {
            return true
        }
        return UIApplication.shared.canOpenURL(URL(string:map.urlPrefix!)!)
    }
}

File: android/src/main/kotlin/com/example/flutter_native_communication/MapHandler.kt
package com.example.flutter_native_communication

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import io.flutter.embedding.engine.plugins.FlutterPlugin


private enum class MapType { google, amap, baidu, waze, yandexNavi, yandexMaps }

private class MapModel(val mapType: MapType, val mapName: String, val packageName: String) {
    fun toMap(): Map<String, String> {
        return mapOf("mapType" to mapType.name, "mapName" to mapName, "packageName" to packageName)
    }
}

class MapHandler(private val context: Context, private val pm: PackageManager) : MethodCallHandler {
    companion object {
        @JvmStatic
        fun registerWith(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
            val channel = MethodChannel(flutterPluginBinding.getBinaryMessenger(), "maps_channel")
            val context = flutterPluginBinding.getApplicationContext()
            val pm = context.getPackageManager()

            channel.setMethodCallHandler(MapHandler(context, pm))
        }
    }

    private val maps = listOf(
            MapModel(MapType.google, "Google Maps", "com.google.android.apps.maps"),
            MapModel(MapType.amap, "Amap", "com.autonavi.minimap"),
            MapModel(MapType.baidu, "Baidu Maps", "com.baidu.BaiduMap"),
            MapModel(MapType.waze, "Waze", "com.waze"),
            MapModel(MapType.yandexNavi, "Yandex Navigator", "ru.yandex.yandexnavi"),
            MapModel(MapType.yandexMaps, "Yandex Maps", "ru.yandex.yandexmaps")
    )

    private fun getInstalledMaps(): List<MapModel> {
        val installedApps = pm.getInstalledApplications(0)
        val installedMaps = maps.filter { map -> installedApps.any { app -> app.packageName == map.packageName } }
        return installedMaps
    }


    private fun isMapAvailable(type: String): Boolean {
        val installedMaps = getInstalledMaps()
        return installedMaps.any { map -> map.mapType.name == type }
    }

    private fun launchGoogleMaps(url: String) {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        intent.setPackage("com.google.android.apps.maps")
        if (intent.resolveActivity(context.packageManager) != null) {
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent)
        }
    }

    private fun launchMap(mapType: MapType, url: String) {
        when (mapType) {
            MapType.google -> launchGoogleMaps(url)
            else -> {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                val foundMap = maps.find { map -> map.mapType == mapType }
                if (foundMap != null) {
                    intent.setPackage(foundMap.packageName)
                }
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                context.startActivity(intent)
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getInstalledMaps" -> {
                val installedMaps = getInstalledMaps()
                result.success(installedMaps.map { map -> map.toMap() })
            }
            "launchMap" -> {
                var args = call.arguments as Map<String, String>

                if (!isMapAvailable(args["mapType"] as String)) {
                    result.error("MAP_NOT_AVAILABLE", "Map is not installed on a device", null)
                    return;
                }

                val mapType = MapType.valueOf(args["mapType"] as String)
                val url = args["url"] as String

                launchMap(mapType, url)
            }
            "isMapAvailable" -> {
                var args = call.arguments as Map<String, String>
                result.success(isMapAvailable(args["mapType"] as String))
            }
            else -> result.notImplemented()
        }
    }
}